File: plep_main/src/deltadak/commands/UndoFacility.java
package deltadak.commands;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Provider for undo operations by keeping commands on a stack.
 */
public class UndoFacility {

    private Deque<Command> undoStack = new ArrayDeque<>();

    /**
     * Saves executed command to possibly revert later.
     * @param command to execute
     */
    public void execute(Command command) {
        command.execute();
        undoStack.push(command);
    }

    /**
     * Undo last done command.
     */
    public void undo() {
        if (!undoStack.isEmpty()) {
            Command command = undoStack.pop();
            if (command.isExecuted()) {
                command.undo();
            }
        }
    }

}

File: src/main/kotlin/nl/deltadak/plep/commands/UndoFacility.kt
package nl.deltadak.plep.commands

import java.util.*

/**
 * Provider for undo operations by keeping commands on a stack.
 */
class UndoFacility {

    private val undoStack = ArrayDeque<Command>()

    /**
     * Saves executed command to possibly revert later.
     * @param command to execute
     */
    fun execute(command: Command) {
        command.execute()
        undoStack.push(command)
    }

    /**
     * Undo last done command.
     */
    fun undo() {
        if (!undoStack.isEmpty()) {
            val command = undoStack.pop()
            if (command.isExecuted) {
                command.undo()
            }
        }
    }

}
